### PR TITLE
ci: Make jemalloc builds reproducible

### PR DESCRIPTION
**Pull request by ptrus**
Also updated development docker image to ubuntu 20.04 so that we can use `ffile-prefix-map` (added in GCC 8)

### CODE REVIEW

File: .changelog/4095.internal.md [hunk 1]
```diff
@@ -0,0 +1 @@
+ci-reproducibility: reproducable jemalloc
```
Comment (tjanez) [inline .changelog/4095.internal.md:1]:
```suggestion
ci: Make jemalloc builds reproducible
```

**Review by tjanez — COMMENTED** [review]
Just tested this locally and if builds are done sequentially, i.e. no `make -j <n-cpus>`, it appears to work.

I'm curious if the following could be an issue:
```
$ diffoscope /tmp/tmp.QP40WBm3re/jemalloc-5.2.1/lib/libjemalloc.a /tmp/tmp.8Ban9KGpxD/jemalloc-5.2.1/lib/libjemalloc.a
 |#                                                                                                                                                            |  N/A%  None  ETA:  --:--:-- --- /tmp/tmp.QP40WBm3re/jemalloc-5.2.1/lib/libjemalloc.a
+++ /tmp/tmp.8Ban9KGpxD/jemalloc-5.2.1/lib/libjemalloc.a
├── file list
│ @@ -1,36 +1,36 @@
│ -----------   0        0        0    10974 2021-06-29 10:00:32.000000 /
│ +----------   0        0        0    10974 2021-06-29 09:59:34.000000 /
│  ----------   0        0        0        0 1970-01-01 00:00:00.000000 //
│ --rw-rw-r--   0     1000     1000  3186304 2021-06-29 10:00:16.000000 jemalloc.o
│ --rw-rw-r--   0     1000     1000  1784808 2021-06-29 10:00:17.000000 arena.o
│ --rw-rw-r--   0     1000     1000   771968 2021-06-29 10:00:18.000000 background_thread.o
│ --rw-rw-r--   0     1000     1000   544504 2021-06-29 10:00:18.000000 base.o
│ --rw-rw-r--   0     1000     1000   462464 2021-06-29 10:00:18.000000 bin.o
│ --rw-rw-r--   0     1000     1000   428160 2021-06-29 10:00:18.000000 bitmap.o
│ --rw-rw-r--   0     1000     1000   910824 2021-06-29 10:00:19.000000 ckh.o
│ --rw-rw-r--   0     1000     1000  3302672 2021-06-29 10:00:22.000000 ctl.o
│ --rw-rw-r--   0     1000     1000   333760 2021-06-29 10:00:22.000000 div.o
│ --rw-rw-r--   0     1000     1000  1363080 2021-06-29 10:00:24.000000 extent.o
│ --rw-rw-r--   0     1000     1000   497600 2021-06-29 10:00:24.000000 extent_dss.o
│ --rw-rw-r--   0     1000     1000   427464 2021-06-29 10:00:24.000000 extent_mmap.o
│ --rw-rw-r--   0     1000     1000   424128 2021-06-29 10:00:24.000000 hash.o
│ --rw-rw-r--   0     1000     1000   436064 2021-06-29 10:00:24.000000 hook.o
│ --rw-rw-r--   0     1000     1000   667096 2021-06-29 10:00:25.000000 large.o
│ --rw-rw-r--   0     1000     1000   433400 2021-06-29 10:00:25.000000 log.o
│ --rw-rw-r--   0     1000     1000   516768 2021-06-29 10:00:25.000000 malloc_io.o
│ --rw-rw-r--   0     1000     1000   481072 2021-06-29 10:00:26.000000 mutex.o
│ --rw-rw-r--   0     1000     1000   446440 2021-06-29 10:00:26.000000 mutex_pool.o
│ --rw-rw-r--   0     1000     1000   434504 2021-06-29 10:00:26.000000 nstime.o
│ --rw-rw-r--   0     1000     1000   504312 2021-06-29 10:00:26.000000 pages.o
│ --rw-rw-r--   0     1000     1000   424128 2021-06-29 10:00:26.000000 prng.o
│ --rw-rw-r--   0     1000     1000   980416 2021-06-29 10:00:27.000000 prof.o
│ --rw-rw-r--   0     1000     1000   494016 2021-06-29 10:00:27.000000 rtree.o
│ --rw-rw-r--   0     1000     1000   427120 2021-06-29 10:00:27.000000 safety_check.o
│ --rw-rw-r--   0     1000     1000  1932784 2021-06-29 10:00:29.000000 stats.o
│ --rw-rw-r--   0     1000     1000   369376 2021-06-29 10:00:30.000000 sc.o
│ --rw-rw-r--   0     1000     1000   350272 2021-06-29 10:00:30.000000 sz.o
│ --rw-rw-r--   0     1000     1000   835488 2021-06-29 10:00:30.000000 tcache.o
│ --rw-rw-r--   0     1000     1000   320760 2021-06-29 10:00:31.000000 test_hooks.o
│ --rw-rw-r--   0     1000     1000   424136 2021-06-29 10:00:31.000000 ticker.o
│ --rw-rw-r--   0     1000     1000   543072 2021-06-29 10:00:31.000000 tsd.o
│ --rw-rw-r--   0     1000     1000   433992 2021-06-29 10:00:31.000000 witness.o
│ --rw-rw-r--   0     1000     1000   551160 2021-06-29 10:00:32.000000 jemalloc_cpp.o
│ +-rw-rw-r--   0     1000     1000  3186304 2021-06-29 09:59:18.000000 jemalloc.o
│ +-rw-rw-r--   0     1000     1000  1784808 2021-06-29 09:59:19.000000 arena.o
│ +-rw-rw-r--   0     1000     1000   771968 2021-06-29 09:59:20.000000 background_thread.o
│ +-rw-rw-r--   0     1000     1000   544504 2021-06-29 09:59:20.000000 base.o
│ +-rw-rw-r--   0     1000     1000   462464 2021-06-29 09:59:20.000000 bin.o
│ +-rw-rw-r--   0     1000     1000   428160 2021-06-29 09:59:20.000000 bitmap.o
│ +-rw-rw-r--   0     1000     1000   910824 2021-06-29 09:59:21.000000 ckh.o
│ +-rw-rw-r--   0     1000     1000  3302672 2021-06-29 09:59:24.000000 ctl.o
│ +-rw-rw-r--   0     1000     1000   333760 2021-06-29 09:59:24.000000 div.o
│ +-rw-rw-r--   0     1000     1000  1363080 2021-06-29 09:59:26.000000 extent.o
│ +-rw-rw-r--   0     1000     1000   497600 2021-06-29 09:59:26.000000 extent_dss.o
│ +-rw-rw-r--   0     1000     1000   427464 2021-06-29 09:59:26.000000 extent_mmap.o
│ +-rw-rw-r--   0     1000     1000   424128 2021-06-29 09:59:26.000000 hash.o
│ +-rw-rw-r--   0     1000     1000   436064 2021-06-29 09:59:27.000000 hook.o
│ +-rw-rw-r--   0     1000     1000   667096 2021-06-29 09:59:27.000000 large.o
│ +-rw-rw-r--   0     1000     1000   433400 2021-06-29 09:59:27.000000 log.o
│ +-rw-rw-r--   0     1000     1000   516768 2021-06-29 09:59:28.000000 malloc_io.o
│ +-rw-rw-r--   0     1000     1000   481072 2021-06-29 09:59:28.000000 mutex.o
│ +-rw-rw-r--   0     1000     1000   446440 2021-06-29 09:59:28.000000 mutex_pool.o
│ +-rw-rw-r--   0     1000     1000   434504 2021-06-29 09:59:28.000000 nstime.o
│ +-rw-rw-r--   0     1000     1000   504312 2021-06-29 09:59:28.000000 pages.o
│ +-rw-rw-r--   0     1000     1000   424128 2021-06-29 09:59:28.000000 prng.o
│ +-rw-rw-r--   0     1000     1000   980416 2021-06-29 09:59:29.000000 prof.o
│ +-rw-rw-r--   0     1000     1000   494016 2021-06-29 09:59:29.000000 rtree.o
│ +-rw-rw-r--   0     1000     1000   427120 2021-06-29 09:59:29.000000 safety_check.o
│ +-rw-rw-r--   0     1000     1000  1932784 2021-06-29 09:59:32.000000 stats.o
│ +-rw-rw-r--   0     1000     1000   369376 2021-06-29 09:59:32.000000 sc.o
│ +-rw-rw-r--   0     1000     1000   350272 2021-06-29 09:59:32.000000 sz.o
│ +-rw-rw-r--   0     1000     1000   835488 2021-06-29 09:59:33.000000 tcache.o
│ +-rw-rw-r--   0     1000     1000   320760 2021-06-29 09:59:33.000000 test_hooks.o
│ +-rw-rw-r--   0     1000     1000   424136 2021-06-29 09:59:33.000000 ticker.o
│ +-rw-rw-r--   0     1000     1000   543072 2021-06-29 09:59:33.000000 tsd.o
│ +-rw-rw-r--   0     1000     1000   433992 2021-06-29 09:59:33.000000 witness.o
│ +-rw-rw-r--   0     1000     1000   551160 2021-06-29 09:59:34.000000 jemalloc_cpp.o
```

It appears only timestamps are different. Can we force timestamps to be the same?

**Comment by ptrus** [comment]
> Can we force timestamps to be the same?

I think this depends on the `ar`/`ranlib` which, at least on ubuntu, defaults to deterministic mode (`-D` flag for both). So i cannot reproduce this. So if you come up with the fix for it (e.g. maybe via `ARFLAGS`?), we can then also explicitly set it here.

**Review by kostko — APPROVED** [review]

**Review by Yawning — APPROVED** [review]

**Comment by tjanez** [comment]
> > Can we force timestamps to be the same?
> 
> I think this depends on the `ar`/`ranlib` which, at least on ubuntu, defaults to deterministic mode (`-D` flag for both). So i cannot reproduce this. So if you come up with the fix for it (e.g. maybe via `ARFLAGS`?), we can then also explicitly set it here.

Indeed, on Fedora, binutils are built with [deterministic archives support turned off by default](https://src.fedoraproject.org/rpms/binutils/blob/fba59c4283a9b56fe85ec173741a3ac54bbe6931/f/binutils.spec#_67).

I haven't found a quick way to fix that (Reproducible Builds project has some details [here](https://reproducible-builds.org/docs/archives/), under _Static libraries_ section) and since this would only impact local developer jemalloc builds, let's leave it as is for now.

**Pull request merged**